test/hotspot/gtest/riscv/test_assembler_riscv.cpp

```diff
@@ -106,4 +106,160 @@ TEST_VM(RiscV, cmov) {
   }
 }
 
+template <typename TESTSIZE, Assembler::operand_size ASMSIZE>
+class CmpxchgTester {
+ public:
+  typedef TESTSIZE (*cmpxchg_func)(intptr_t addr, TESTSIZE expected, TESTSIZE new_value, TESTSIZE result);
+
+  static TESTSIZE base_cmpxchg(int variant, intptr_t addr, TESTSIZE expected, TESTSIZE new_value, TESTSIZE result, bool boolean_result = false) {
+    BufferBlob* bb = BufferBlob::create("riscvTest", 128);
+    CodeBuffer code(bb);
+    MacroAssembler _masm(&code);
+    address entry = _masm.pc();
+    {
+      switch(variant) {
+        default:
+          _masm.cmpxchg(/*addr*/ c_rarg0, /*expected*/ c_rarg1, /*new_value*/c_rarg2,
+                        ASMSIZE, Assembler::relaxed, Assembler::relaxed,
+                        /*result*/ c_rarg3, boolean_result);
+          _masm.mv(c_rarg0, c_rarg3);
+          break;
+        case 1:
+          // expected == result
+          _masm.cmpxchg(/*addr*/ c_rarg0, /*expected*/ c_rarg1, /*new_value*/c_rarg2,
+                        ASMSIZE, Assembler::relaxed, Assembler::relaxed,
+                        /*result*/ c_rarg1, boolean_result);
+          _masm.mv(c_rarg0, c_rarg1);
+          break;
+        case 2:
+          // new_value == result
+          _masm.cmpxchg(/*addr*/ c_rarg0, /*expected*/ c_rarg1, /*new_value*/c_rarg2,
+                        ASMSIZE, Assembler::relaxed, Assembler::relaxed,
+                        /*result*/ c_rarg2, boolean_result);
+          _masm.mv(c_rarg0, c_rarg2);
+          break;
+        case 3:
+          // expected == new_value
+          _masm.cmpxchg(/*addr*/ c_rarg0, /*expected*/ c_rarg1, /*new_value*/ c_rarg1,
+                        ASMSIZE, Assembler::relaxed, Assembler::relaxed,
+                        /*result*/ c_rarg2, boolean_result);
+          _masm.mv(c_rarg0, c_rarg2);
+          break;
+
+      }
+      _masm.ret();
+    }
+    _masm.flush();
+    OrderAccess::cross_modify_fence();
```

Comment:
Question: do we need this `cross_modify_fence`? Seems to me the preceding `_masm.flush()` will do here. Same question for the `CmovTester` case.

Reply:
The exact implementation of flush we may want to change, e.g. Zjid, I/D consistency.
And the same goes for CMF, e.g. with Zijd it may be an IMPORT.I and FENCE.I or similar.
Here we don't care about that, we only want to test some instructions.

The scary options here is UseCtxFencei, which allows us to avoid some icache flush.
And we may want to remove more icache flushes.

By having both we get a pretty good guarantee that future optimizations don't break this gtest.
As producer I call flush(), as consumer I call CMF.

But yes we can remove CMF, I'm fine with that also, maybe my paranoia is not warranted.

Reply:
I see. But I am a bit afraid that this could be misleading to people. We have the CMF, but we only have one thread working as both producer and consumer here. Maybe we can keep it simple for now and update as appropriate when those cases you mentioned happen?

Reply:
Sure